bot/src/export.rs

```diff
@@ -0,0 +1,638 @@
+
+        // Baseline stake amount: ◎5000.000000000
+        v.push((vec![
+            "baseline_stake_amount",
+        ], Regex::new(r"Baseline stake amount: ◎([\d\\.]+)").unwrap()));
+
+        // Bonus stake amount: ◎46269.599541788
+        v.push((vec![
+            "bonus_stake_amount",
+        ], Regex::new(r"Bonus stake amount: ◎([\d\\.]+)").unwrap()));
+
+        // Validators by stake level: None=200, Baseline=177, Bonus=1659
+        v.push((vec![
+            "num_no_stake_validators",
+            "num_baseline_stake_validators",
+            "num_bonus_stake_validators",
+        ], Regex::new(r"Validators by stake level: None=(\d+), Baseline=(\d+), Bonus=(\d+)").unwrap()));
+
+        v.push((vec![
+            "min_testnet_participation_numerator",
+            "min_testnet_participation_denominator",
+        ], Regex::new(r"Participants must maintain Baseline or Bonus stake level for (\d+) of the last (\d+) Testnet epochs").unwrap()) );
+
+        v
+    };
+}
+
+/// parse the notes in the yaml file to get the validator program epoch data
+fn read_epoch_notes(notes: &Vec<String>) -> Map<String, Value> {
+    let mut note_vals = Map::new();
+
+    for note in notes {
+        let mut captures;
+
+        let mut found = false;
+
+        for (vars, regex) in NOTE_REGEXES.iter() {
+            captures = regex.captures(&note);
+            if captures.is_some() {
+                found = true;
+                let cap = captures.unwrap();
+
+                for (i, var) in vars.iter().enumerate() {
+                    let str_value = &cap[i + 1];
+
+                    let maybe_number: Result<Number, _> = serde_json::from_str(str_value);
+
+                    let value: Value;
+                    if maybe_number.is_ok() { // didn't parse as number; make string
+                        value = serde_json::to_value(maybe_number.unwrap()).unwrap();
+                    } else {
+                        value = serde_json::to_value(str_value).unwrap();
+                    }
+                    note_vals.insert(var.to_string(), value);
+                };
+
+                break;
+            }
+        }
+        if !found {
+            warn!("Could not find match for: {}", note);
+        }
+    }
+
+    note_vals
+}
+
+fn persist_epoch_notes(
+    notes: &Map<String, Value>,
+    transaction: &mut Transaction<'_>,
+    config: &Config,
+    epoch: Epoch,
+) -> BoxResult<()> {
+
+    // usize does not have toSql trait, so change to i32, which does
+    let epoch_i32 = i32::try_from(epoch)?;
+    let rows = transaction.query("SELECT id from \"EpochStats\" WHERE epoch=$1", &[&epoch_i32])?;
+
+    let num_rows = rows.len();
+    if num_rows == 1 {
+        let id: i32 = rows.first().unwrap().get("id");
+        info!("UPDATE EpochStats row with id {}", id);
+
+        transaction.execute(
+            "UPDATE \"EpochStats\" \
```

Comment:
Could use a raw string to clean up the escaping here, eg: https://rahul-thakoor.github.io/rust-raw-string-literals/